### PR TITLE
Environment Pytorch Dataset

### LINKED ISSUE #16: Pytorch DataLoader class. Data API for feeding the networks
A pytorch dataloader is required to sample data and forward it to the networks.
`class EnvironmentSampler(Dataset):`
The class will have the following structure:

- A constructor: `def __init__(self, environment, length):`
where `environment` is an `Environment` instance to indicate the physical system to draw data from; and `length` indicates the length of the dataset (since we can keep sampling random rollouts we avoid having an infinite size).

- `    def __len__(self):
        return self.length`
- An iterable call `def __getitem__(self, i):`
this will be called each time a batch is required. A rollout is sampled and converted as a pytorch tensor to be forwarded to the network. It should perform standarization of the data (maybe just mean substraction?)

 Sample code will be provided in the same file to show its usage.

### PR DESCRIPTION
**Pull request by charlio23**
EnvironmentSampler Pytorch DataLoader for data sampling.
Samples rollouts from an environment, performs standarization, returns pytorch array in the following format (Batch, Nframes, Channels, Height, Width).

### CODE REVIEW

File: environments/datasets.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        return rolls.transpose((0, 1, 4, 2, 3))
+
+
+# Sample code for DataLoader call
+if __name__ == "__main__":
+    import time
+    pd = Pendulum(mass=.5, length=1, g=3)
+    trainDS = EnvironmentSampler(
+        environment=pd,
+        length=100,
```
Comment (OleguerCanal) [inline environments/datasets.py:83]: Dont call it length also, can be confused with the length of the pendulum

File: environments/datasets.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    def __len__(self):
+        """Get dataset length
+
+        Returns:
+            length (int): Length of the dataset.
+        """
+        return self.length
+
+    def __getitem__(self, i):
```
Comment (OleguerCanal) [inline environments/datasets.py:54]: what does it return? Is it a Torch.tensor?

File: environments/datasets.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            number_of_rollouts (int): Number of rollouts to generate.
+            img_size (int): Size of the frames (in pixels).
+            color (bool): Whether to have colored or grayscale frames.
+            noise_std (float): Standard deviation of the gaussian noise source, no noise for noise_std=0.
+            radius_bound (float, float): Radius lower and upper bound of the phase state sampling.
+                Init phase states will be sampled from a circle (q, p) of radius
+                r ~ U(radius_bound[0], radius_bound[1]) https://arxiv.org/pdf/1909.13789.pdf (Sec. 4)
+            world_size (float) Spatial extent of the window where the rendering is taking place (in meters).
+            seed (int): Seed for reproducibility.
+            data_mean (float): Data mean for standaritzation. Defaults to 0.5.
```
Comment (OleguerCanal) [inline environments/datasets.py:30]: Shouldnt this default to 0 and std dev to 1 to have normalized data?

File: environments/datasets.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        rolls = self.environment.sample_random_rollouts(number_of_frames=self.number_of_frames,
+                                                        delta_time=self.delta_time,
+                                                        number_of_rollouts=self.number_of_rollouts,
+                                                        img_size=self.img_size,
+                                                        noise_std=self.noise_std,
+                                                        radius_bound=self.radius_bound,
+                                                        world_size=self.world_size,
+                                                        seed=self.seed)
+        # standarization
+        rolls = (rolls - self.data_mean)/self.data_std
```
Comment (OleguerCanal) [inline environments/datasets.py:73]: Arent the rolls from 0-255? I don't understand how you are normalizing here

File: environments/datasets.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        self.number_of_frames = number_of_frames
+        self.delta_time = delta_time
+        self.number_of_rollouts = number_of_rollouts
+        self.img_size = img_size
+        self.noise_std = noise_std
+        self.radius_bound = radius_bound
+        self.world_size = world_size
+        self.seed = seed
+
+    def __len__(self):
```
Comment (OleguerCanal) [inline environments/datasets.py:46]: is this needed?

**Review by OleguerCanal — CHANGES_REQUESTED** [review]
Nice work! Just minor comments :)

File: environments/datasets.py (hunk 3 above)
Comment (charlio23) [reply environments/datasets.py:30]: If we have an image that is [0...1] to have standarized data [-1 ... 1] we can use 0.5 as mean and std, that is why i took this default value

File: environments/datasets.py (hunk 4 above)
Comment (charlio23) [reply environments/datasets.py:73]: the rolls are [0 ... 1]

File: environments/datasets.py (hunk 5 above)
Comment (charlio23) [reply environments/datasets.py:46]: yes, since EnvironmentSampler class is treated as an iterable in the training loop. I have not tested what happens if this methods is missing but i am sure it won't work without it.

File: environments/datasets.py (hunk 2 above)
Comment (charlio23) [reply environments/datasets.py:54]: yes, i forgot!

**Comment by charlio23** [comment]
I solved the issues. I left some conversations unresolved cuz I am not sure if they answer your doubts.

File: environments/datasets.py (hunk 3 above)
Comment (OleguerCanal) [reply environments/datasets.py:30]: Oh i see, soory. I thought the images were 0-255
